icicle/tests/test_hash_api.cpp

```diff
@@ -0,0 +1,215 @@
+}
+
+/****************************** Merkle **********************************/
+class HashSumBackend : public HashBackend
+{
+public:
+  HashSumBackend(uint64_t total_input_limbs, uint64_t total_output_limbs, uint64_t total_secondary_input_limbs = 0)
+      : HashBackend(total_input_limbs, total_output_limbs, total_secondary_input_limbs)
+  {
+  }
+
+  virtual eIcicleError hash_single(
+    const limb_t* input_limbs,
+    limb_t* output_limbs,
+    const HashConfig& config,
+    const limb_t* secondary_input_limbs = nullptr) const
+  {
+    output_limbs[0] = 0;
+    for (int i = 0; i < m_total_input_limbs; ++i) {
+      output_limbs[0] += input_limbs[i];
+    }
+    for (int i = 1; i < m_total_output_limbs; ++i) {
+      output_limbs[i] += output_limbs[0];
+    }
+    return eIcicleError::SUCCESS;
+  }
+
+  virtual eIcicleError hash_many(
+    const limb_t* input_limbs,
+    limb_t* output_limbs,
+    int nof_hashes,
+    const HashConfig& config,
+    const limb_t* secondary_input_limbs = nullptr) const
+  {
+    for (int i = 0; i < nof_hashes; ++i) {
+      hash_single(input_limbs, output_limbs, config, secondary_input_limbs);
+      input_limbs += m_total_input_limbs;
+      output_limbs += m_total_output_limbs;
+      secondary_input_limbs += m_total_secondary_input_limbs;
+    }
+    return eIcicleError::SUCCESS;
+  }
+};
+
+Hash create_hash_sum(uint64_t total_input_limbs, uint64_t total_output_limbs, uint64_t total_secondary_input_limbs = 0)
+{
+  auto backend = std::make_shared<HashSumBackend>(total_input_limbs, total_output_limbs, total_secondary_input_limbs);
+  return Hash(backend);
+}
+
+TEST_F(HashApiTest, MerkleTree)
+{
+  // const uint64_t nof_leaves_limbs = 100; // Number of input limbs
+  // auto leaves = std::make_unique<uint32_t[]>(nof_leaves_limbs);
+  // // Randomize the input array
+  // randomize(leaves.get(), nof_leaves_limbs);
+
+  auto config = default_merkle_tree_config();
+  auto layer0_hash = create_hash_sum(5, 2); // input 5 limbs, output 2 limbs
+  auto layer1_hash = create_hash_sum(6, 3); // input 6 limbs, output 3 limbs
+  auto layer2_hash = create_hash_sum(6, 2); // input 6 limbs, output 2 limbs
+  auto merkle_tree =
+    create_merkle_tree({layer0_hash, layer1_hash, layer2_hash}, 1 /*limbs per leaf*/, 2 /*min level to store*/);
+
+  limb_t leaves[100];
+  for (int i = 0; i < 100; ++i) {
+    leaves[i] = i;
+  }
+  // std::cout << “build start” << std::endl;
+  merkle_tree.build(leaves, config);
+  // merkle_tree.print_tree();
+  // const limb_t* root;
+  // merkle_tree.get_root(root);
+  // std::cout << "Root = " << *root << std::endl;
+  // limb_t* path;
+  // uint path_size;
+  // uint64_t element_idx = 5;
+  // merkle_tree.allocate_path(path, path_size);
+  // merkle_tree.get_path(leaves, element_idx, path, config);
+  // std::cout << “path(“<< element_idx << “):” << std::endl;
+  // merkle_tree.print_path(path);
+  // std::cout << “verify:” << std::endl;
```

Comment:
shouldn't these lines be uncommented? otherwise, the test only checks that the tree doesn't crash upon building it

Reply:
yes it's WIP